src/chain/sync/grandpa_warp_sync.rs

```diff
@@ -0,0 +1,364 @@
+    }
+
+    /// Add a peer to the list of peers.
+    pub fn add_peer(mut self, peer: PeerId) -> GrandpaWarpSync {
+        self.peers.push(peer);
+
+        GrandpaWarpSync::WarpSyncRequest(WarpSyncRequest {
+            peer_index: self.peer_index,
+            peers: self.peers,
+            genesis_chain_information: self.genesis_chain_information,
+        })
+    }
+
+    /// Remove a peer from the list of peers.
+    ///
+    /// Will panic if `to_remove` is the `current_peer`.
+    pub fn remove_other_peer(mut self, to_remove: PeerId) -> GrandpaWarpSync {
+        if to_remove == self.current_peer() {
+            panic!("Atttempted to remove the current peer.");
+        }
+
+        let index = self.peers.iter().position(|peer| peer == &to_remove);
+
+        if let Some(index) = index {
+            // There's no point in removing a peer if it's behind the current index.
+            if index > self.peer_index {
+                self.peers.remove(index);
+            }
+        }
+
+        GrandpaWarpSync::WarpSyncRequest(WarpSyncRequest {
+            peer_index: self.peer_index,
+            peers: self.peers,
+            genesis_chain_information: self.genesis_chain_information,
+        })
+    }
+
+    /// Submit a GrandPa warp sync network response if the request succeeded or
+    /// `None` if it did not.
+    pub async fn handle_response(
+        self,
+        response: Option<Vec<GrandpaWarpSyncResponseFragment>>,
+    ) -> GrandpaWarpSync {
+        let warp_sync_peer = self.current_peer();
+
+        let next_index = self.peer_index + 1;
+
+        match response {
+            Some(response_fragments) => GrandpaWarpSync::Verifier(Verifier {
+                verifier: warp_sync::Verifier::new(
+                    &self.genesis_chain_information,
+                    response_fragments,
+                ),
+                genesis_chain_information: self.genesis_chain_information,
+                warp_sync_peer,
+            }),
+            None if next_index < self.peers.len() => GrandpaWarpSync::WarpSyncRequest(Self {
+                peer_index: next_index,
+                peers: self.peers,
+                genesis_chain_information: self.genesis_chain_information,
+            }),
+            None => GrandpaWarpSync::WaitingForPeers(WaitingForPeers {
+                peers: Vec::new(),
+                genesis_chain_information: self.genesis_chain_information,
+            }),
+        }
+    }
+}
+
+/// Fetching the parameters for the virtual machine is required to continue.
+pub struct VirtualMachineParamsGet {
+    state: PostVerificationState,
+}
+
+impl VirtualMachineParamsGet {
+    /// Set the code and heappages from storage using the keys `:code` and `:heappages`
+    /// respectively. Also allows setting an execution hint for the virtual machine.
+    pub fn inject_virtual_machine_params(
+        self,
+        code: impl AsRef<[u8]>,
+        heap_pages: u64,
```

Comment:
```suggestion
        code: Option<impl AsRef<[u8]>>,
        heap_pages: Option<impl AsRef<[u8]>>,
```

In order to make it more convenient for the user, I think you should parse `code` and `heap_pages` directly here.

If either is `None`, then return `Finished` with an error.
`heap_pages` should be parsed using `u64::from_le_bytes`